### PR TITLE
Refactor MDK to use pluggable runtime

### PR DESCRIPTION
**Pull request by itamarst**
MINOR BACKWARDS INCOMPATIBILITY: Some constructors (but _not_ the publicly documented MDK-level API) now take an extra argument. Usage via the handy static factory methods is still the same. So long as services using internal APIs are pinned to particular MDK version this shouldn't impact anyone, and the necessary minor change can be done when needed. @kflynn @plombardi89 

Changes:
1. Added dependency injection system.
2. Added the world's tiniest actor system, with a no-reentrancy guarantee.
3. Added external runtime, currently falling back to Quark runtime. At the moment this just implements time and scheduling, but now it has actual tests that are run against both the real runtime and against a fake double for writing tests with.
4. Modified `WSClient` to use the new scheduling actor.
5. A copy of Quark's `Promise` class is included, slightly modified to integrate with the actor system ensure-no-reentrancy message delivery system. Once we switch to using this class we can drop the `Promise` in Quark.

### CODE REVIEW

File: quark/actors_core.q [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	void startActor(Actor actor) {
+	    self._queue(new _StartActor(actor, self));
+	}
+
+	@doc("Queue a message for delivery.")
+	void _queue(_QueuedMessage inFlight) {
+	    self._lock.acquire();
+	    self._queued.add(inFlight);
+	    self._lock.release();
+	    if (self._delivering) {
```
Comment (bozzzzo) [inline quark/actors_core.q:86]: Looking at self._delivering outside of self._lock. Copy to a local var inside lock.

File: quark/mdk_runtime.q [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    An actor that can schedule events.
+
+    Accepts Schedule messages and send Happening events to originator at the
+    appropriate time.
+    """)
+    interface SchedulingActor extends Actor {}
+
+    @doc("""
+    Please send me a Happening message with given event name in given number of
+    milliseconds.
```
Comment (bozzzzo) [inline quark/mdk_runtime.q]: float milliseconds?

File: quark/mdk_runtime.q [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	    _scheduled[_scheduled.keys().size()] = new _FakeTimeRequest(origin, sched.event, self._now + sched.seconds);
+	}
+
+	float time() {
+	    return self._now;
+	}
+
+	@doc("Run scheduled events whose time has come.")
+	void pump() {
+	    long idx = 0;
```
Comment (bozzzzo) [inline quark/mdk_runtime.q]: `int idx = 0` ?

quark complains a bit with https://travis-ci.org/datawire/mdk/builds/148381548#L503

or is this the future-ready plan on having just one  integral type?

Comment (itamarst) [inline quark/mdk_runtime.q]: future-readiness, yeah. Although I guess we'd probably want to call the combo type `int`, cause what's a `long`? ;) Can change if you think it's wrong way to do it.

File: quark/mdk_runtime.q (hunk 2 above)
Comment (itamarst) [inline quark/mdk_runtime.q]: Oops, will fix.

File: quark/mdk_runtime.q (hunk 3 above)
Comment (bozzzzo) [inline quark/mdk_runtime.q]: I think it's kind-of nice if MDK compiles without quark warnings ;) (until we kill the warnings)

**Comment by rhs** [comment]
I took a look and didn't see any reason not to land this assuming it works as advertised. I'd say it's probably worth doing an end-to-end test in ruby and java prior to landing since I don't think our native tests actually cover those languages (correct me if I'm wrong).

The one question that popped to mind from reading the code is about shutdown. Looking at the Actor.onStart() makes me wonder about the possibility of Actor.onStop() which makes me wonder about our shutdown story in general (avoiding stack traces/hanging/etc). Probably beyond the scope of this pull request, but perhaps worth noting somewhere as a requirement to address going forward?

**Comment by itamarst** [comment]
Also should increase vesrion number on all the files I changed.

**Pull request merged**